[JSC] Optimize `Array#of` in DFG/FTL

sosukesuzuki opened the pull request:
#### 445df9a64416167ce85f555cfc7827aa5446282d
<pre>
[JSC] Optimize `Array#of` in DFG/FTL
<a href="https://bugs.webkit.org/show_bug.cgi?id=301681">https://bugs.webkit.org/show_bug.cgi?id=301681</a>

Reviewed by Keith Miller.

This patch changes to optimize `Array#of` in DFG/FTL using existing `NewArray` node.

                                   TipOfTree                  Patched

array-of-contiguous              2.1078+-0.1328     ^      1.6572+-0.1088        ^ definitely 1.2719x faster
array-of-double                  1.5172+-0.0460     ^      0.7300+-0.0452        ^ definitely 2.0784x faster
array-of-int32-large            18.9503+-0.1974     ^      7.0067+-0.5997        ^ definitely 2.7046x faster
array-of-double-large           23.7362+-0.9485     ^      7.3903+-0.7021        ^ definitely 3.2118x faster
array-of-contiguous-large       21.9875+-2.7947     ^     10.6635+-0.5827        ^ definitely 2.0619x faster
array-of                         1.3079+-0.0263     ^      0.6680+-0.0389        ^ definitely 1.9581x faster

* JSTests/microbenchmarks/array-of-contiguous-large.js:
* JSTests/microbenchmarks/array-of-contiguous.js:
* JSTests/microbenchmarks/array-of-double-large.js:
* JSTests/microbenchmarks/array-of-double.js:
* JSTests/microbenchmarks/array-of-int32-large.js:
* JSTests/microbenchmarks/array-of.js:
* JSTests/stress/array-of-contiguous.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-double.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-int32.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-non-array-constructor.js: Added.
(shouldBe):
(NonArray):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::ArrayConstructor::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:

Canonical link: <a href="https://commits.webkit.org/302565@main">https://commits.webkit.org/302565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f104db09af6b63fa33a4c7ea1582703845e1f344

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1764 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/40346 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136890 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/80935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1696 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1640 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/98646 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/80935 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/132454 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/1317 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115998 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79293 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/1238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34129 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80164 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/121499 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/109708 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34627 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/139364 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/127959 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1554 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/1489 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/107172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1596 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112339 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107013 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27247 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1268 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30857 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/54246 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1625 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64988 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/160973 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/1445 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/40152 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/1479 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1547 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/runtime/Intrinsic.h:
```diff
@@ -211,6 +211,7 @@ namespace JSC {
     macro(PromiseRejectIntrinsic) \
     macro(PromisePrototypeThenIntrinsic) \
     macro(PromisePrototypeCatchIntrinsic) \
+    macro(ArrayConstructorOfIntrinsic) \
```
Nit: Might be worth putting this after the `Array.prototype` intrinsics.

kmiller68 reviewed (approved):
r=me with nit

sosukesuzuki committed:
[JSC] Optimize `Array#of` in DFG/FTL
https://bugs.webkit.org/show_bug.cgi?id=301681

Reviewed by Keith Miller.

This patch changes to optimize `Array#of` in DFG/FTL using existing `NewArray` node.

                                   TipOfTree                  Patched

array-of-contiguous              2.1078+-0.1328     ^      1.6572+-0.1088        ^ definitely 1.2719x faster
array-of-double                  1.5172+-0.0460     ^      0.7300+-0.0452        ^ definitely 2.0784x faster
array-of-int32-large            18.9503+-0.1974     ^      7.0067+-0.5997        ^ definitely 2.7046x faster
array-of-double-large           23.7362+-0.9485     ^      7.3903+-0.7021        ^ definitely 3.2118x faster
array-of-contiguous-large       21.9875+-2.7947     ^     10.6635+-0.5827        ^ definitely 2.0619x faster
array-of                         1.3079+-0.0263     ^      0.6680+-0.0389        ^ definitely 1.9581x faster

* JSTests/microbenchmarks/array-of-contiguous-large.js:
* JSTests/microbenchmarks/array-of-contiguous.js:
* JSTests/microbenchmarks/array-of-double-large.js:
* JSTests/microbenchmarks/array-of-double.js:
* JSTests/microbenchmarks/array-of-int32-large.js:
* JSTests/microbenchmarks/array-of.js:
* JSTests/stress/array-of-contiguous.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-double.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-int32.js:
(throw.new.Error): Deleted.
* JSTests/stress/array-of-non-array-constructor.js: Added.
(shouldBe):
(NonArray):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::handleIntrinsicCall):
* Source/JavaScriptCore/runtime/ArrayConstructor.cpp:
(JSC::ArrayConstructor::finishCreation):
* Source/JavaScriptCore/runtime/Intrinsic.h:

Canonical link: https://commits.webkit.org/302565@main
